test/cpp/qps/client_async.cc

```diff
@@ -248,6 +248,17 @@ class AsyncClient : public ClientImpl<StubType, RequestType> {
     ClientRpcContext* ctx = ClientRpcContext::detag(got_tag);
     std::mutex* shutdown_mu = &shutdown_state_[thread_idx]->mutex;
     shutdown_mu->lock();
+    if (shutdown_state_[thread_idx]->shutdown) {
+      ctx->TryCancel();
+      delete ctx;
+      while (cli_cqs_[cq_[thread_idx]]->Next(&got_tag, &ok)) {
+        ctx = ClientRpcContext::detag(got_tag);
+        ctx->TryCancel();
```

Comment:
is this a sync operation?

Reply:
I don't think so, the doc says this is "out of band".

FWIW, this is the same shutdown logic that has been used for a while.  This just adds a shutdown check before the main loop, the issue was that things were getting shut down before we entered the main loop, and we didn't check first.

Reply:
Nice catch. Since this is just duplicated code from down below, can you take it out and make it a separate function or lambda or something?

Reply:
Fixed

Reply:
I ask because it looks scary to delete the ctx right after an async op has been performed "on" it. But if this is kosher, LGTM

Reply:
dgq to answer your question: TryCancel just happens purely asynchronously without any CQ tag returning or anything, so it's fine to do a TryCancel and then delete the ctx.